src/parser/ast/ForInOfStatementNode.h

```diff
@@ -0,0 +1,170 @@
+ *  USA
+ */
+
+#ifndef ForInOfStatementNode_h
+#define ForInOfStatementNode_h
+
+#include "ExpressionNode.h"
+#include "StatementNode.h"
+
+namespace Escargot {
+
+class ForInOfStatementNode : public StatementNode {
+public:
+    friend class ScriptParser;
+    ForInOfStatementNode(Node *left, Node *right, Node *body, bool forIn)
+        : StatementNode()
+        , m_left(left)
+        , m_right(right)
+        , m_body((StatementNode *)body)
+        , m_forIn(forIn)
+    {
+    }
+
+    virtual ~ForInOfStatementNode()
+    {
+    }
+
+    virtual ASTNodeType type()
+    {
+        if (m_forIn) {
+            return ASTNodeType::ForInStatement;
+        } else {
+            return ASTNodeType::ForOfStatement;
+        }
+    }
+
+    virtual void generateStatementByteCode(ByteCodeBlock *codeBlock, ByteCodeGenerateContext *context)
+    {
+        bool canSkipCopyToRegisterBefore = context->m_canSkipCopyToRegister;
+        context->m_canSkipCopyToRegister = false;
+
+        ByteCodeGenerateContext newContext(*context);
+
+        newContext.getRegister(); // ExecutionResult of m_right should not overwrite any reserved value
+        if (m_left->type() == ASTNodeType::VariableDeclaration)
+            m_left->generateResultNotRequiredExpressionByteCode(codeBlock, &newContext);
+
+        size_t baseCountBefore = newContext.m_registerStack->size();
+        size_t rightIdx = m_right->getRegister(codeBlock, &newContext);
+        m_right->generateExpressionByteCode(codeBlock, &newContext, rightIdx);
+        codeBlock->pushCode(LoadLiteral(ByteCodeLOC(m_loc.index), newContext.getRegister(), Value()), &newContext, this);
+        size_t literalIdx = newContext.getLastRegisterIndex();
+        newContext.giveUpRegister();
+        size_t equalResultIndex = newContext.getRegister();
+        newContext.giveUpRegister();
+        codeBlock->pushCode(BinaryEqual(ByteCodeLOC(m_loc.index), literalIdx, rightIdx, equalResultIndex), &newContext, this);
+        codeBlock->pushCode(JumpIfTrue(ByteCodeLOC(m_loc.index), equalResultIndex), &newContext, this);
+        size_t exit1Pos = codeBlock->lastCodePosition<JumpIfTrue>();
+
+        codeBlock->pushCode(LoadLiteral(ByteCodeLOC(m_loc.index), literalIdx, Value(Value::Null)), &newContext, this);
+        codeBlock->pushCode(BinaryEqual(ByteCodeLOC(m_loc.index), literalIdx, rightIdx, equalResultIndex), &newContext, this);
+        codeBlock->pushCode(JumpIfTrue(ByteCodeLOC(m_loc.index), equalResultIndex), &newContext, this);
+        size_t exit2Pos = codeBlock->lastCodePosition<JumpIfTrue>();
+
+        size_t exit3Pos, continuePosition;
+        if (m_forIn) {
+            // for-in statement
+            size_t ePosition = codeBlock->currentCodeSize();
+            codeBlock->pushCode(EnumerateObject(ByteCodeLOC(m_loc.index)), &newContext, this);
+            codeBlock->peekCode<EnumerateObject>(ePosition)->m_objectRegisterIndex = rightIdx;
```

Comment:
How about instead of peeking the last pushed code, simply pass the `rightIdx` to the `EnumerateObject` constructor? (Same as for the for-of statement)

Reply:
Thank you for the nice suggestion. But, other necessary member variable, `m_dataRegisterIndex`, of `EnumerateObject` cannot be defined when the `EnumerateObject` constructor called, so I set each member variable after the constructor called. Basically, (to my knowledge) each constructor of ByteCode initializes member variables when all of its values are defined.

Reply:
Okay, that's reasonable, thanks for the explanation.